=====
Usage
=====

The tool starts from an interaction matrix and can therefore be combined with any tool that maps, filters and normalizes Hi-C data to generate a matrix. In this case, BWA, HICUP and Hi-C Pro were used to generate interaciton matrices from sequencing data.

HiCpro2homer

This tool converts (iced) Hi-C matrices generated by HiC-Pro into the format of Homer. While the homer matrix reports both directions of an interaction (a -> b and b -> a), the Hi-C pro format is less redundant as it stores each interaction value of two bins under one ID. The Hi-C pro format also skips the report of zero value interactions.

Hi-C Matrix

HiCMatrix first tests the shape of the converted matrix. To be able to compare different iced matrices (with different column sums), normalize_by_columns_sum allows to divide each column by the column sum median, thereby turning the interaction values into frequencies. In some cases, it may be useful to restrict the analysis to a subset of the matrix (select). Thereby, distinct chromosomes can be in- (keep_pattern) or excluded (remove_pattern) from the overall matrix, resulting in a submatrix that can be saved separately or used instead of the original matrix for the following analyses (inplace).
Hi-C Matrix also gives the possibility to generate differential matrices by dividing the matrix (or submatrix) by a denominator matrix. As some bins contain zero values, a pseudocount of choice can be added to each bin of the nominator and denominator matrix (default: 0.001).

Heatmap
HiC
The subcommand heatmap prints matrices, submatrices or differential matrices of choice, either as whole genome heatmaps (_plot_heatmap_globally) or chromosome-wise (_plot_heatmap_split_by_chrom; two bins required as minimum lenght). The heatmap can either be plotted in a square format or as triangles (--rotate). Colors are based on the colour palettes of seaborn and can be adjusted as desired (default: sns.cubehelix_palette for non-differential heatmaps and Rdblue for differential heatmaps) and the minimum and maximum values of the color scale can be adjusted (._vim, ._vmax). The output file (_write_heatmap_to_file) can either be in png or pdf format. In png format, all generated heatmaps are printed in separate files, whereas in pdf format, all files are printed to a single file. Also, the png file resolution can be set as desired (._png_dpi).

Colocalization

The colocalization tester allows to select regions of interest and to test them for increased interchromosomal interaction frequencies in two ways.  Either, genomic loci of interest can be selected by keywords/features from the corresponding genome gff file (._feature), which are then translated into the corresponding bins of the interaction matrix and collected. Alternatively, genomic regions of any length and choice can be subjected to the colocalization tester in gff format (.read_gff_file), and overlapping bins will be compiled from the matrix (.read_matrix_file_and_add_as_features) in the same way. The regions of interest will be tested against a randomized set of bins (.extract_random_bins), that mirrors the query sample in distribution and length of spanned bins. The distributions of query and randomized sample are plotted (.plot_distribution) and tested for significant differences (.perform_t_test) using Welch’s t-test statistics. This sampling can be redone as many times as desired (._number_of_subsamplings) and bins with zero values can be in- or excluded from the analysis (._remove zeros). Additionally, the tool allows to add various numbers of flanking bins to the regions of interest (._margin) and also to restrict analysis to these (._flanks_only). As output, the tool gives the individual interaction counts from the query and the randomized sample, a histogram of those as a pdf, as well as mean and median values with according standard deviation and p-values.

Diff_colocalization

Differential colocalization can be used to compare distinct interchromosomal interactions in two different matrices. It assesses the interaction frequencies of genomic loci of interest over a randomized background sample (as described in Colocalization) in two matrices of choice and tests for significant differences in signal-over-background among them. In each round of subsampling, the query sample is tested against a randomized background sample in both matrices. The ratio between average signal and average background is compiled and the subsampling is repeated (as often as desired), thus resulting in two respective populations of signal-over-background ratios (one for each matrix). The difference between these two populations is tested using Welch’s t-test statistics.

Virtual 4C

This tool enables to use any region of choice from the full-genome HiC matrix as bait for a virtual 4C. The regions of choice can be submitted in gff format (see example: Chromosome – start – end). Subsequently, these genomic regions are translated into the corresponding bins of the HiC matrix and the affected columns are extracted. If several columns were selected, an average interaction value for every genomic bin is calculated. The interaction values for each bin are translated into a wiggle file, where step and window size correspond to the resolution of the input matrix. 

Distance Dependent Decay

The Distance Dependent Decay calculates bin-wise the mean or median interaction frequencies for any given genomic distance. This can be done for individual chromosomes or across multiple chromosomes. Further, the analysis can be conducted arm-wise by splitting the individual chromosomes at the site of the centromere (requires positional information about the centromeres).
